### ISSUE TITLE
Multithreading!

### ISSUE THREAD

**Issue by jamestwebber** [czbiohub/bcl2fastr#40]
"the fun part"

Once we have the samplesheet parsing working correctly (#39) it is time to break up the decoding/writing into multiple threads, each with a subset of the samples to work on. Once that is working we can start to profile performance on different size datasets / samplesheets.

 - [x] Update the arguments to allow for varying n_threads
 - [ ] Figure out how multi-threading in Rust works, and divvy up the samples to different threads that run in parallel
 - [x] This is probably the right time to think about any profiling tooling we need

**Comment by jamestwebber** [comment]
I did a little of this in the `index_count` branch for #63. For multithreading I used `rayon` which is almost a drop-in replacement for existing iterators. Our case is fairly simple because we don't need any message passing or other locks, so a simple `par_iter` worked well.

For profiling I tried [`flamegraph`](https://github.com/ferrous-systems/flamegraph/) which worked pretty well. In the context of multithreading it's a bit hard to interpret the results, but it helped show me that a lot of time was spent in `alloc` and `free` for e.g. the filters.
